src/library/analysisfeature.cpp

```diff
@@ -112,88 +111,74 @@ void AnalysisFeature::activate() {
     emit(enableCoverArtDisplay(true));
 }
 
-namespace {
-    inline
-    AnalyzerQueue::Mode getAnalyzerQueueMode(
-            const UserSettingsPointer& pConfig) {
-        if (pConfig->getValue<bool>(ConfigKey("[Library]", "EnableWaveformGenerationWithAnalysis"), true)) {
-            return AnalyzerQueue::Mode::Default;
-        } else {
-            return AnalyzerQueue::Mode::WithoutWaveform;
-        }
-    }
-} // anonymous namespace
-
 void AnalysisFeature::analyzeTracks(QList<TrackId> trackIds) {
-    if (m_pAnalyzerQueue == NULL) {
+    if (!m_pAnalyzerQueue) {
         // Save the old BPM detection prefs setting (on or off)
         m_iOldBpmEnabled = m_pConfig->getValueString(ConfigKey("[BPM]","BPMDetectionEnabled")).toInt();
         // Force BPM detection to be on.
         m_pConfig->set(ConfigKey("[BPM]","BPMDetectionEnabled"), ConfigValue(1));
         // Note: this sucks... we should refactor the prefs/analyzer to fix this hacky bit ^^^^.
 
-        m_pAnalyzerQueue = new AnalyzerQueue(
-                m_pDbConnectionPool,
-                m_pConfig,
-                getAnalyzerQueueMode(m_pConfig));
+        m_pAnalyzerQueue = AnalyzerQueuePointer(
+                m_library,
+                kNumberOfAnalyzerThreads,
```

Comment:
This is the other problem that I told you. Since you don't have a single manager, but instead two Queues, one in AnaliserFeature and the other in PlayerManager, you cannot control that the number of threads is at much idealThreadCount.
My implementation has a single manager, and has a mechanism (that is not perfect, but worked) that reduces the number of threads dedicated to batch analysis while the player analyses a loaded track.

Reply:
I would like to avoid such a mechanism that is dynamically rescheduling work between threads at run time. Why don't we advise users not to run a batch analysis while performing? I even remember that batch analysis in Serato was only enabled in offline mode with no hardware connected. Not each and every problem must be solved by coding. Sometimes organizational solutions are much more effective and help to reduce the complexity of the code.

As @daschuer mentioned idle threads don't consume many resources except their stack frame and some heap space allocated by the analyzers. The number of threads should not be an issue if they are not running concurrently.

And I didn't want to combine the two thread pools (let's call them what they actually are) now to avoid even more disrupting changes in the surrounding code.

Reply:
In my opinion, your changes are more disrupting than my PR was, so that should not be a factor.

If you take a look at my fork, it does not reschedule threads, but instead pauses some of the batch processing threads to give processing power to the player threads.

If you don't want to make it as complicated as I did (those lists that I have are not nice), you could pause all batch threads while a player analysis is ongoing. That would be a better solution for the user than just having to actually stop the analysis. (So, you'd need some signals from playermanager to analysisfeature)

Reply:
> pause all batch threads while a player analysis is ongoing

👍